### PR TITLE
display and CRUD for experiment channels

### PR DESCRIPTION
**Pull request by snopoke**
* Display channels on the experiment details page
* Ability to add, update and delete channels

Peek 2023-10-17 17-13

### CODE REVIEW

File: apps/channels/admin.py [hunk 1]
```diff
@@ -30,7 +31,3 @@ def team(self, obj):
 
     def get_changeform_initial_data(self, request: HttpRequest) -> Dict[str, str]:
         return {"extra_data": {"bot_token": "your token here"}}
```
Comment (SmittieC) [inline apps/channels/admin.py:33]: Out of curiosity, what was the thinking of including them again?

File: apps/channels/models.py [hunk 2]
```diff
@@ -12,16 +12,42 @@
...
+            cls.WHATSAPP,
+        ]
+
+    def form(self):
+        from apps.channels.forms import ChannelForm
+
+        return ChannelForm(initial={"platform": self})
+
+    def extra_form(self, *args, **kwargs):
+        from apps.channels.forms import TelegramChannelForm, WhatsappChannelForm
```
Comment (SmittieC) [inline apps/channels/models.py:33]: Can this be moved to the top of the file, or were there circular import errors?

Comment (snopoke) [reply apps/channels/models.py:33]: It can't because forms imports from models

File: apps/experiments/views/experiment.py [hunk 3]
```diff
@@ -145,17 +145,60 @@ def single_experiment_home(request, team_slug: str, experiment_id: int):
...
+        ExperimentChannel, id=channel_id, experiment_id=experiment_id, experiment__team__slug=team_slug
+    )
+    if request.POST.get("action") == "delete":
+        channel.delete()
+    else:
+        form = channel.form(data=request.POST)
+        if form.is_valid():
+            extra_form = channel.extra_form(data=request.POST)
+            config_data = {}
+            if extra_form and extra_form.is_valid():
```
Comment (SmittieC) [inline apps/experiments/views/experiment.py:196]: would `extra_form` not always exist here?

File: apps/channels/admin.py (hunk 1 above)
Comment (snopoke) [reply apps/channels/admin.py:33]: This could actually be removed now

**Review by SmittieC — APPROVED** [review]

**Pull request merged**